### PR TITLE
Add underscore formatting in numbers

### LINKED ISSUE #7015: Number Formatting
### Suggestion

Basically, make it so you can specify numbers with underscores.
Example: `set {_n} to 10_000`, `give player 1_024 of barrier`

### Why?

1) Most programming languages, such as Python and Java, already have this feature.
2) For big numbers, it is a pain to write them out and count the length to make sure the number is correct and that we didn't enter a high number.

### Other

_No response_

### Agreement

- [x] I have read the guidelines above and affirm I am following them with this suggestion.

### PR DESCRIPTION
**Pull request by Efnilite**
### Description
Resolves #7015. Adds underscore support for formatting in numbers. 

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** #7015

### CODE REVIEW

**Comment by Fusezion** [comment]
I don't believe we need a regression test currently but an actual test added, this isn't so much resolving a bug as much as adding an enhancement to numbers.

**Comment by Efnilite** [comment]
> I don't believe we need a regression test currently but an actual test added, this isn't so much resolving a bug as much as adding an enhancement to numbers.

Should I put it in the misc folder then?

**Comment by Fusezion** [comment]
Yeah I think misc is fair as this isn't syntax related

**Comment by Asleeepp** [comment]
Kenzie mentioned something about adding a way to detect if it was a valid number, something like ____1___ shouldn't be valid

**Comment by UnderscoreTud** [comment]
> Kenzie mentioned something about adding a way to detect if it was a valid number, something like ____1___ shouldn't be valid
> 
> heres the discord conversation if you're curious: [here](https://discord.com/channels/135877399391764480/836220422223036467/1279182736682254376)

That already isn't valid, but my main concern is how you can have adjacent underscores, for example `1___0` is valid when I don't think it makes that much sense.

**Comment by Efnilite** [comment]
> Kenzie mentioned something about adding a way to detect if it was a valid number, something like ____1___ shouldn't be valid
> 
> heres the discord conversation if you're curious: [here](https://discord.com/channels/135877399391764480/836220422223036467/1279182736682254376)

This has already been addressed in the last lines of the test. Parsing the numbers starting and/or ending with `_` seems to resolve to none, as these numbers don't match the updated regex. If anything else should be changed, please lmk

**Comment by Efnilite** [comment]
> > Kenzie mentioned something about adding a way to detect if it was a valid number, something like __**1**_ shouldn't be valid
> > heres the discord conversation if you're curious: [here](https://discord.com/channels/135877399391764480/836220422223036467/1279182736682254376)
> 
> That already isn't valid, but my main concern is how you can have adjacent underscores, for example `1___0` is valid when I don't think it makes that much sense.

Personally I agree that having multiple `_` is a little silly but someone might have a genuine use case for it, like maybe someone prefers `1__000__000` over `1_000_000`. Java also support this. 

I do think adding this is beneficial, although it would be possible to limit it to just one `_` if more people think it's a good idea.

**Comment by Asleeepp** [comment]
> > > Kenzie mentioned something about adding a way to detect if it was a valid number, something like __**1**_ shouldn't be valid
> > > heres the discord conversation if you're curious: [here](https://discord.com/channels/135877399391764480/836220422223036467/1279182736682254376)
> > 
> > 
> > That already isn't valid, but my main concern is how you can have adjacent underscores, for example `1___0` is valid when I don't think it makes that much sense.
> 
> Personally I agree that having multiple `_` is a little silly but someone might have a genuine use case for it, like maybe someone prefers `1__000__000` over `1_000_000`. Java also support this.
> 
> I don't think this reason should axe this entire PR. It would be possible to limit it to just one `_` if more people think it's a good idea.

then i think that it would be better to make it consistent across the whole number
someone cant do `1__000_000` for example
though maybe there should also be a limit on how many underscores you can use

**Comment by UnderscoreTud** [comment]
> > > Kenzie mentioned something about adding a way to detect if it was a valid number, something like __**1**_ shouldn't be valid
> > > heres the discord conversation if you're curious: [here](https://discord.com/channels/135877399391764480/836220422223036467/1279182736682254376)
> > 
> > 
> > That already isn't valid, but my main concern is how you can have adjacent underscores, for example `1___0` is valid when I don't think it makes that much sense.
> 
> Personally I agree that having multiple `_` is a little silly but someone might have a genuine use case for it, like maybe someone prefers `1__000__000` over `1_000_000`. Java also support this.
> 
> I don't think this reason should axe this entire PR. It would be possible to limit it to just one `_` if more people think it's a good idea.

Me and @APickledWalrus talked about it on discord and he also thought the same thing (that it shouldn't be allowed), but we can wait and see what the other members think

**Comment by sovdeeth** [comment]
oh yes i definitely don't think it you should axe this pr
but i do agree that only 1 underscore makes sense and should be the requirement

**Comment by Efnilite** [comment]
Aight, changed it to at most one and added tests for making sure that having multiple isn't supported

File: src/test/skript/tests/misc/underscores in numbers.sk [hunk 1]
```diff
@@ -8,33 +8,36 @@ test "underscores in numbers":
...
-	assert ("1.2__3_6___" parsed as number) is not a number with "1.2__3_6___ is a number"
+	assert ("1.2_3_6___" parsed as number) is not a number with "1.2_3_6___ is a number"
 
 	assert ("_1_" parsed as number) is not a number with "_1_ is a number"
-	assert ("_1.2__3_6_" parsed as number) is not a number with "_1.2__3_6_ is a number"
+	assert ("_1.2_3_6_" parsed as number) is not a number with "_1.2_3_6_ is a number"
 	assert ("___1___" parsed as number) is not a number with "___1___ is a number"
-	assert ("___1.2__3_6___" parsed as number) is not a number with "___1.2__3_6___ is a number"
\ No newline at end of file
+	assert ("___1.2_3_6___" parsed as number) is not a number with "___1.2_3_6___ is a number"
```
Comment (Asleeepp) [inline src/test/skript/tests/misc/underscores in numbers.sk:43]:
```suggestion
	assert ("___1.2_3_6___" parsed as number) is not a number with "___1.2_3_6___ is a number"

```

**Review by Asleeepp — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

**Review by Asleeepp — APPROVED** [review]
since you requested my review (my approval does nothing haha)

**Review by Moderocky — APPROVED** [review]

**Pull request merged**